dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/lifecycle/ApplicationStartBarrierImpl.java

```diff
@@ -0,0 +1,111 @@
+
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.wsspi.application.lifecycle.ApplicationPrereq;
+import com.ibm.wsspi.application.lifecycle.ApplicationStartBarrier;
+
+/**
+ * Enable Declarative Services to enforce start sequencing.
+ * Application handlers should depend on this service.
+ * Services should implement {@link ApplicationPrereq}
+ */
+@Component(configurationPolicy = ConfigurationPolicy.REQUIRE)
+public class ApplicationStartBarrierImpl implements ApplicationStartBarrier {
+    private static final TraceComponent tc = Tr.register(ApplicationStartBarrierImpl.class);
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE)
+    void setApplicationPrereq(ApplicationPrereq applicationPrereq) {
+    }
+
+    /**
+     * Non configurable immediate component to track prereqs as they appear.
+     */
+    @Component(configurationPolicy = ConfigurationPolicy.IGNORE, immediate = true)
+    public static class PrereqWatcher {
+        private final ConfigurationAdmin configurationAdmin;
+        private final ArrayList<String> realisedPrereqs = new ArrayList<>();
+
+        /**
+         * Use constructor parameter references to ensure that mandatory references are supplied first.
+         *
+         * @param configurationAdmin must be supplied before any calls to setApplicationPrereq
+         */
+        @Activate
+        public PrereqWatcher(@Reference ConfigurationAdmin configurationAdmin) {
+            this.configurationAdmin = configurationAdmin;
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "Known configured application prereqs:" + listConfiguredPrereqs());
+            }
+        }
+
+        private String listConfiguredPrereqs() {
+            Configuration[] configs;
+            try {
+                configs = this.configurationAdmin.listConfigurations("(service.factoryPid=" + ApplicationPrereq.class.getName() + ")");
+            } catch (IOException | InvalidSyntaxException e) {
+                StringWriter sw = new StringWriter();
+                e.printStackTrace(new PrintWriter(sw));
+                return "Error listing prereqs:" + sw;
+            }
+            if (configs == null)
+                return null;
+
+            String[] prereqClassNames = new String[configs.length];
+            for (int i = 0; i < configs.length; i++) {
+                prereqClassNames[i] = (String) configs[i].getProperties().get("className");
+            }
+
+            Arrays.sort(prereqClassNames);
+            return Arrays.toString(prereqClassNames);
+        }
+
+        @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
+        void setApplicationPrereq(ApplicationPrereq applicationPrereq) {
+            realisedPrereqs.add(applicationPrereq.getClass().getName());
+            Collections.sort(realisedPrereqs);
```

Comment:
Why always sort even when trace is not enabled.  I only see realisedPrereqs getting used within trace blocks.

Reply:
Very good question. Would it be acceptable to use a sorted set instead, or should we simply defer sorting until we know it needs to be used?

Reply:
I suppose a sorted set would be the most simple.  The number of prereqs likely is not high so sorting as you add likely should not show any measurable performance difference.